### ISSUE TITLE
Ungenerated libs.version.toml file in dependency prevents addition to composite build

### ISSUE THREAD

**Issue by MatthewFoster624** [gradle/gradle#19288]
### Expected Behavior
Expected behavior for composite builds would be that gradle sub-projects already built successfully can be added to a composite build. 
If the _libs.version.toml_ file is required (or needs rebuilt) for the composite build to complete, it can be generated automatically during the build, since it is accomplished manually using `./gradlew generateCatalogsAsToml`.

### Current Behavior
Composite build fails with an error message of: 
```
org.gradle.api.InvalidUserDataException: Invalid catalog definition:
  - Problem: In version catalog [myDependencyNameHere], import of external catalog file failed.
 ```
Full stack trace reveals the issue is a missing `libs.version.toml` file.

### Context
Issue resulted from trying to run a composite build consisting of 10+ projects, all of which built fine on their own.
Running the composite build, however, repeatedly produced the error shown above. This required figuring out which project was failing, loading it, running the `generateCatalogsAsToml`, then rebuilding it before retrying the composite build.

### Steps to Reproduce 
Composite project pseudocode is: 
```
for (project in projects) {
    if (project exists) {
          includeBuild(project)
   } else {
       //skips nonexistent projects
   }
}
```
To reproduce with certainty, remove the `libs.version.toml` file from one of the projects to be included in the composite build, then try to build the composite.

### Your Environment
Gradle 7.2

**Comment by anderso** [comment]
I have encountered the same issue and have made a minimal reproducible case that uses gradle 7.6 at:

https://github.com/anderso/gradle-reproduce-composite-catalog

It contains three separate gradle projects:

  - catalog: a project using the version-catalog and maven-publish plugins to define and publish a catalog as described in https://docs.gradle.org/current/userguide/platforms.html#sec:version-catalog-plugin
  - project: a project that uses the version catalog defined above
  - composite: a composite build containing the above projects

To reproduce

```
git clone https://github.com/anderso/gradle-reproduce-composite-catalog.gi
cd gradle-reproduce-composite-catalog/composite/
./gradlew build
```
Produces:

```
FAILURE: Build failed with an exception.

* What went wrong:
org.gradle.api.InvalidUserDataException: Invalid catalog definition:
  - Problem: In version catalog libs, import of external catalog file failed.
    
    Reason: File '/tmp/gradle-reproduce-composite-catalog/catalog/build/version-catalog/libs.versions.toml' doesn't exist.
    
    Possible solution: Make sure that the catalog file 'libs.versions.toml' exists before importing it.
    
    Please refer to https://docs.gradle.org/7.6/userguide/version_catalog_problems.html#catalog_file_does_not_exist for more details about this problem.
> Invalid catalog definition:
    - Problem: In version catalog libs, import of external catalog file failed.
      
      Reason: File '/tmp/gradle-reproduce-composite-catalog/catalog/build/version-catalog/libs.versions.toml' doesn't exist.
      
      Possible solution: Make sure that the catalog file 'libs.versions.toml' exists before importing it.
      
      Please refer to https://docs.gradle.org/7.6/userguide/version_catalog_problems.html#catalog_file_does_not_exist for more details about this problem.
```

So the problem is that "project" expects the toml file to available in the build-directory of the catalog project.

This is annoying because it means that we cannot have a version catalog in a composite build that works the same way as other projects do. I guess this is a chicken and egg problem, because the initialization phase of "project" in this case depends on the output of "catalog", so the composite project cannot be initialized. One possible solution would be if when we use a libs.version.toml as input to version-catalog-plugin, then we could shortcircuit the plugin altogether in this case, so that "project" could use the toml file directly from the source directory instead of the output of the build.

The workaround is:

```
cd ../catalog
./gradlew generateCatalogAsToml
```
Then the composite project can be built. I'm not sure if this consititutes a bug but it sure would be nice if this could be made to work. In any case I'm not sure if the current behavior is helpful, because it seems that this usecase is halfway supported and undocumented, any changes in the catalog will not propagate until you execute generateCatalogAsToml again, or publish which depends on generateCatalogAsToml.

**Comment by jjohannes** [comment]
> I guess this is a chicken and egg problem, because the initialization phase of "project" in this case depends on the output of "catalog", so the composite project cannot be initialized. 

That's correct. And it's an unsolved issue since the feature to publish and import a catalog was introduced.

I think this can't really be solved when you define the version catalog in the `build.gradle` file as the evaluation of all `settings.gradle` files in a composite, which contain the catalog import, is always done before build files are even looked at.

The currently best solution in my mind is to define the catalog in TOML format and publish that directly. Then you avoid the generation step. For this you need to reconfigure the _outgoing variant/configuration_ to use a TOML file as artifact instead of the result of the generation task:

```
configurations.versionCatalogElements {
    outgoing.artifacts.clear()
    outgoing.artifact(layout.projectDirectory.dir("gradle/libs.versions.toml"))
}
```

And then remove the catalog definition from `build.gradle` and put it in `gradle/libs.versions.toml` (or a toml file in another location) instead.

I think this is a fine solution and maybe the `version-catalog` plugin could offer an option to configure this out of the box.

**Comment by anderso** [comment]
@jjohannes Thanks! I implemented the fix in the test case and can confirm that it works as intended. The composite build picks up the version directly from the toml file. And by changing the catalog extension to use the file:

```
catalog {
    versionCatalog {
        from(files("gradle/libs.versions.toml"))
    }
}
```
It is also published correctly by the version catalog plugin. So this does exactly what I want. 

I added the fix to the branch "workaround" in the test case: https://github.com/anderso/gradle-reproduce-composite-catalog/commit/beca220
